### ISSUE TITLE
Footer is too often included on non-English websites

### ISSUE THREAD

**Issue by bittlingmayer** [mozilla/readability#468]
It seems like removing the footer is done by matching English-specific text.

**Comment by bittlingmayer** [comment]
Concretely we see problems for major publishers in Russian, Ukrainian, Greek, Armenian, Czech, German and Catalan.

**Comment by gijsk** [comment]
Removing a lot of stuff (and selecting which bits to keep) is done by matching classes/ids of elements, see https://github.com/mozilla/readability/blob/076bf2017be0e28df6a46042eee52209a477f5e6/Readability.js#L111-L127 . We could potentially update those regexes for the pages in question, or if you have alternative suggestions how to do better without relying on the language used for the markup/structure of the page, I'd be interested... Is there an obvious alternative?

Can you post some example links of the pages you're having issues with?

**Comment by bittlingmayer** [comment]
+ @minas90

**Comment by bittlingmayer** [comment]
> Is there an obvious alternative?

I would not want to cause regressions for English.  I am biased but I think long-term the solution should be learned from data ie statistical.

Short-term I think looking for HTML tags is more scalable than looking for words in each language.  It's just that in each market there are a few publishing platforms so the tags differ slightly but you do get a lot of coverage with just a few.

And just to keep the code clean and make it easy for people to contribute I would factor out the English and German that is already there into some data file or at least a constant somewhere in this file.

```
const PREV = [
  // en
  'continue', 'next',
  // de
  'weiter'
  // ru
  ...
]
```
And then `prevLink = new RegExp(PREV.join("|"), 'gi');`

**Comment by gijsk** [comment]
>  I am biased but I think long-term the solution should be learned from data ie statistical.

That requires properly tagged/processed examples to "learn" from... I think it might be quite tricky to find the right way to feed the data in, in that case, and/or to have so many examples that it doesn't matter... In any case, as you already implied, it would be a very very big project to convert Readability to something based on statistics / machine learning.

> Short-term I think looking for HTML tags is more scalable than looking for words in each language. 

I think you're right that we should do more with `<main>` and `<article>` tags than we do, but I think they are also still abused regularly, so I'm not sure if this is really likely to be sufficient. Typical NYT markup still includes a container with a pile of `<div>` elements where some are ads (and match the JS regex `/\bad\b/` in the element ID or class) and others are article paragraphs. The NYT is sadly far from alone in this. So I'm skeptical about this being sufficient.

> It's just that in each market there are a few publishing platforms so the tags differ slightly but you do get a lot of coverage with just a few.

This is good to know...

> And just to keep the code clean and make it easy for people to contribute I would factor out the English and German that is already there into some data file or at least a constant somewhere in this file.

I'd take a patch that moved the current alphabetized regexes into a more structured form that we process into regexes. We have reasonable test coverage to avoid regressing anything major immediately, so that should help.

I can see the appeal to having another file, but at the moment I think the amount of data is still small, and the size of Readability is also not terrible yet, so I think I'd keep it in Readability.js for now. If its size gets out of hand we can reconsider. The problem with separating it out completely is that then you end up needing a build system or `require()` support or whatever, and that makes things less portable than they are now...

**Comment by bittlingmayer** [comment]
I was going to give the more structured form a try, but I cannot find `prevLink` and `nextLink` actually used anywhere.  Could it be that they are unused?

**Comment by gijsk** [comment]
Yes, I think we ripped out support for pagination - it used to be in a pile of commented out code, and I must have missed that these properties are unused.

**Comment by bittlingmayer** [comment]
So footers are mostly in `negative`?

**Comment by gijsk** [comment]
Yeah, that and `unlikelyCandidates`, I expect.
